components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/persistence/cache/CacheBackedPolicyDAO.java

```diff
@@ -0,0 +1,178 @@
+        return policies;
+    }
+
+    @Override
+    public void deletePAPPolicy(String policyId, int tenantId) throws EntitlementException {
+
+        super.deletePAPPolicy(policyId, tenantId);
+        papPolicyCache.clearCacheEntry(policyId, tenantId);
+        papPolicyListCache.clearCacheEntry(tenantId, tenantId);
+    }
+
+    @Override
+    public PolicyStoreDTO getPDPPolicy(String policyId, int tenantId) {
+
+        PolicyStoreDTO policy = pdpPolicyCache.getValueFromCache(policyId, tenantId);
+        if (policy != null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("Cache hit in PdpPolicyCache for policy: %s for tenant: %s",
+                        policyId, tenantId));
+            }
+            return policy;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Cache miss in PdpPolicyCache for policy: %s for tenant: %s", policyId, tenantId));
+        }
+        policy = super.getPDPPolicy(policyId, tenantId);
+        pdpPolicyCache.addToCache(policyId, policy, tenantId);
+        pdpPolicyListCache.clearCacheEntry(tenantId, tenantId);
+        return policy;
+    }
+
+    @Override
+    public PolicyStoreDTO[] getAllPDPPolicies(int tenantId) throws EntitlementException {
+
+        PolicyStoreDTO[] policies = pdpPolicyListCache.getValueFromCache(tenantId, tenantId);
+        if (policies != null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("Cache hit in PdpPolicyListCache for policies for tenant: %s", tenantId));
+            }
+            return policies;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Cache miss in PdpPolicyListCache for policies for tenant: %s", tenantId));
+        }
+        policies = super.getAllPDPPolicies(tenantId);
+        pdpPolicyListCache.addToCache(tenantId, policies, tenantId);
+        return policies;
+    }
+
+    @Override
+    public void insertOrUpdatePolicy(PolicyStoreDTO policy, int tenantId) throws EntitlementException {
+
+        super.insertOrUpdatePolicy(policy, tenantId);
+        pdpPolicyCache.addToCache(policy.getPolicyId(), policy, tenantId);
+        pdpPolicyListCache.clearCacheEntry(tenantId, tenantId);
+    }
+
+    @Override
+    public void updateActiveStatusAndOrder(PolicyStoreDTO policy, int tenantId) throws EntitlementException {
+
+        super.updateActiveStatusAndOrder(policy, tenantId);
+        pdpPolicyCache.addToCache(policy.getPolicyId(), policy, tenantId);
+        pdpPolicyListCache.clearCacheEntry(tenantId, tenantId);
+    }
+
+    @Override
+    public int getPublishedVersion(PolicyStoreDTO policy, int tenantId) throws EntitlementException {
+
+        String policyId = policy.getPolicyId();
+        PolicyStoreDTO cachedPolicy = pdpPolicyCache.getValueFromCache(policyId, tenantId);
+        if (cachedPolicy != null && StringUtils.isNotBlank(cachedPolicy.getVersion())) {
```

Comment:
Why `StringUtils.isNotBlank(cachedPolicy.getVersion())` part required here?

Reply:
added this because `getPDPPolicy` methods do not expect the version to be set in thePolicyStoreDTO object (we do set it currently when retrieving, but it appears in registry impl it does not). So there may be cases where we won't set this value, hence handled here.

Reply:
But registry impl doesn't not interfere with caching here.. right?

Reply:
Yes, in that case will remove the additional check